### PR TITLE
Fix expanding structs in queries with ORDER BY

### LINKED ISSUE #9433: INTERNAL Error: Attempted to access index 1 within vector of size 1
### What happens?

When attempting to use an `ORDER BY` clause in a query, I get the below error. Without the `ORDER BY` clause, there is no issue with the query and it executes successfully.

After the error is received, the database is in a bad state/crashed, and I must re-connect to run more queries.

Query:
```
SELECT UNNEST(player_statistics) FROM 
	(SELECT player.id AS player_id,
         UNNEST (player_stats.statistics).team.id AS team_id,
         UNNEST (player_stats.statistics).games.appearences * UNNEST (player_stats.statistics).games.minutes AS totals
	 FROM player_stats) AS player_statistics
WHERE team_id = 48
AND totals > 0
ORDER BY totals
```

```
org.jkiss.dbeaver.model.sql.DBSQLException: SQL Error: java.sql.SQLException: INTERNAL Error: Attempted to access index 1 within vector of size 1
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:133)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:583)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$1(SQLQueryJob.java:492)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:190)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:499)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:947)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:4100)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:123)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:190)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:121)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:5164)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.sql.SQLException: java.sql.SQLException: INTERNAL Error: Attempted to access index 1 within vector of size 1
	at org.duckdb.DuckDBPreparedStatement.prepare(DuckDBPreparedStatement.java:112)
	at org.duckdb.DuckDBPreparedStatement.execute(DuckDBPreparedStatement.java:183)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:330)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:131)
	... 12 more
Caused by: java.sql.SQLException: INTERNAL Error: Attempted to access index 1 within vector of size 1
	at org.duckdb.DuckDBNative.duckdb_jdbc_prepare(Native Method)
	at org.duckdb.DuckDBPreparedStatement.prepare(DuckDBPreparedStatement.java:106)
	... 15 more
```

### To Reproduce

1. Download the following duckdb database file: https://github.com/jbaranski/duck-db-bug-report/blob/main/data.db
2. Connect to the database in DBeaver (I used `Version 23.2.3.202310221533`) on MacOS `Ventura 13.4.1` with an `M1` chip.
3. Run the query successfully:
```
SELECT UNNEST(player_statistics) FROM 
	(SELECT player.id AS player_id,
         UNNEST (player_stats.statistics).team.id AS team_id,
         UNNEST (player_stats.statistics).games.appearences * UNNEST (player_stats.statistics).games.minutes AS totals
	 FROM player_stats) AS player_statistics
WHERE team_id = 48
AND totals > 0
```
4. Run the same query with an `ORDER BY` clause, it breaks:
```
SELECT UNNEST(player_statistics) FROM 
	(SELECT player.id AS player_id,
         UNNEST (player_stats.statistics).team.id AS team_id,
         UNNEST (player_stats.statistics).games.appearences * UNNEST (player_stats.statistics).games.minutes AS totals
	 FROM player_stats) AS player_statistics
WHERE team_id = 48
AND totals > 0
ORDER BY totals
```

### OS:

MacOS - Ventura 13.4.1

### DuckDB Version:

0.9.1

### DuckDB Client:

Java/DBeaver

### Full Name:

Jeff Baranski

### Affiliation:

Hobby Project / No Affiliation

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by taniabogatsch**
In `Binder::BindSelectNode`, we call `BindModifierTypes`. This is where we throw the assertion or where we access the index out of bounds. 

Before this, we already bind the ORDER BY expression. However, we never add the `result_type` of that expression to the `modifier_sql_types` vector because we `continue` the for-loop before that. This causes the out-of-bounds exception. Simply fixing this by adding the type to the vector before calling `continue` does not solve the problem, as the ORDER BY is bound to a 'wrong' column index if we expand structs (e.g., with UNNEST).

This fix introduces a `modifier_expansion_count` to track how many select list expressions we expand, i.e., how much our column index shifts (as we increase the select list). With this, we can adjust the ORDER BY `binding` to access the correct `sql_type` and have the correct column index.

This fixes #9433.

### CODE REVIEW

File: src/planner/binder/query_node/bind_select_node.cpp [hunk 1]
```diff
@@ -445,15 +460,25 @@ unique_ptr<BoundQueryNode> Binder::BindSelectNode(SelectNode &statement, unique_
...
 			if (!is_original_column) {
 				throw InternalException("Only original columns can have expanded expressions");
 			}
 			if (statement.aggregate_handling == AggregateHandling::FORCE_AGGREGATES) {
 				throw BinderException("UNNEST of struct cannot be combined with GROUP BY ALL");
 			}
+
 			auto &struct_expressions = select_binder.ExpandedExpressions();
 			D_ASSERT(!struct_expressions.empty());
+			modifier_expansion_count.back() += struct_expressions.size() - 1;
```
Comment (Mytherin) [inline src/planner/binder/query_node/bind_select_node.cpp:480]: Couldn't we instead also push the types of the struct expansion into `internal_sql_types` here so they are available while binding the ORDER BY?

File: test/sql/order/test_order_by.test [hunk 2]
```diff
@@ -176,3 +176,17 @@ SELECT a-10 AS k FROM test UNION SELECT a-11 AS l FROM test ORDER BY a-11;
...
+# issue #9433
+
+statement ok
+CREATE TABLE tbl AS SELECT {'a': 2.0, 'b': 'hello', 'c': [1, 2]} AS s;
+
+statement ok
+INSERT INTO tbl VALUES ({'a': 1.0, 'b': 'yay', 'c': [10, 20]});
+
+query IIII
+SELECT UNNEST(s), s.a AS id FROM tbl ORDER BY id;
```
Comment (Mytherin) [inline test/sql/order/test_order_by.test:188]: What would happen if we were to refer to the unnest in the ORDER BY, e.g.

```sql
SELECT UNNEST(s), s.a AS id FROM tbl ORDER BY 1;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Comments:

File: src/planner/binder/query_node/bind_select_node.cpp (hunk 1 above)
Comment (taniabogatsch) [reply src/planner/binder/query_node/bind_select_node.cpp:480]: I tried that, but we still don't know how much the column index of the ORDER BY column changes due to the expansion, so we can't get the correct type later if we don't know the shift (I think). `modifier_expansion_count` is to keep track of that. And to update the column index correctly. If I follow the code correctly, we first bind the ORDER BY modifiers, so their column indexes don't account for any potential expansions.

Comment (taniabogatsch) [reply src/planner/binder/query_node/bind_select_node.cpp:480]: But I understand the concern of pushing the dummy type. I will look more into it.

Comment (Mytherin) [reply src/planner/binder/query_node/bind_select_node.cpp:480]: Perhaps we could do the expansion in an earlier stage instead, which is what we do for e.g. `*` expressions?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**